### ISSUE TITLE
Why cannot we generate shortIDS with os.urandom and base64?

### ISSUE THREAD

**Issue by Natim** [corpix/shortid#1]
``` python
>>> import os, base64
>>> base64.b64encode(os.urandom(6))
'TKmj0DI4'
```

**Comment by ghost** [comment]
To get a response regarding what is special about this approach to generating an ID, you might want to post this where it was created: https://github.com/dylang/shortid

**Comment by Natim** [comment]
The version you've pointed out is in nodejs, here I am just fixing the Python implementation. However this solution might also works with Node :+1:

**Comment by ghost** [comment]
This library is a port of an ID generation strategy that was originally published in NodeJS. Your question is essentially *Why generate the ID this way, when you can do it simpler this way?"

I would be careful about considering using the standard Math library random() function as the basis for a unique ID. I am not a crypto specialist or anything, but I have read enough articles about the theory to know that the standard random() functions in most languages do not generate a high degree of randomness. For a _table_ id, it probably is good enough, but I get the feeling that some of the techniques used in ShortID are to support better randomness as well as improved randomness within clusters.

If you want to have the author of this strategy explain why it works the way it does and how it compares to just using random() with base64 encoding, I would ask dylang.

FYI - at least one difference I can see between what you propose and how ShortID works now is that your strategy produces a fixed length ID (example above - 8 characters). ShortID produces an ID between 7 and 14 characters, and the longer you generate ID's in ShortID (tens of thousands), it grows longer over time. Is that significant for a _table_ scoped ID - probably not :)

**Comment by Natim** [comment]
> I have read enough articles about the theory to know that the standard random() functions in most languages do not generate a high degree of randomness.

Yes that's exactly why I would like to us the urandom function instead.

**Comment by Natim** [comment]
> ShortID produces an ID between 7 and 14 characters,

This is really strange, you should be able to define the number of bytes you want for your scope with regards to your collection size but we could add a random from 5 to 9 bytes.

**Comment by ghost** [comment]
Sorry, I am not familiar with Python, but I see now that _os_ implies operating system, so it seems that what you are recommending is the Linux urandom function? This (crazy long) article http://www.2uo.de/myths-about-urandom/ says that urandom is a valid crypto-based random generator, but I suppose this means it only works on Linux?

**Comment by Natim** [comment]
No Python `os.urandom` method has bindings for all operating systems.

**Comment by ghost** [comment]
I am going to reiterate my original point and sign off this discussion. This was never meant to be a critique of the method you are proposing (which I am sure is fine). It was meant to highlight the fact that this repo (which I have nothing to do with) is a port of a specific ID generation strategy that was authored somewhere else (https://github.com/dylang/shortid).

**Comment by Natim** [comment]
Yes I have got that thanks :) I will try to get it addredded there as well.

**Comment by Natim** [comment]
I did some math and here is what I found: https://github.com/dylang/shortid/issues/60#issuecomment-190651009

**Comment by Natim** [comment]
For the record here is a piece of Python script that I use to generate shortIDs:
It is based on cryptographic random bytes.

```
import base64
import os
import random


def shortid(nb_bytes=6):
    return base64.urlsafe_b64encode(
        os.urandom(nb_bytes)).decode('utf-8').rstrip('=')


class ShortId(object):
    # From 7 to 14 characters: Up to 64^7 + 64^8 + ... + 64^14
    def __init__(self, min_bytes=5, max_bytes=10):
        self.min_bytes = min_bytes
        self.max_bytes = max_bytes

    def generate(self, nb_bytes=None):
        if not nb_bytes:
            random.seed = os.urandom(1024)
            nb_bytes = random.randint(self.min_bytes, self.max_bytes)
        return shortid(nb_bytes)
```

**Comment by corpix** [comment]
Thank you for the interest in my fork! I'll dive into your changes and discussion in a few days, sorry don't have enough time. At this moment I think that this is a good change, but not sure about the base64.

**Comment by Natim** [comment]
Hello @corpix base64 is just another name for the alphabet you are using. It is also easier to reason about because each base64 character handle a 6bit value.

**Comment by corpix** [comment]
Sorry for taking so much time to answer.
Thank you for this great thread, I am starting to think that solution that involves `os.urandom` is more reliable. But I don't like an idea to use base64 because of the extra characters it requires, I know, there is `altchars` kwarg, but I think that more clean way is to implement encoding from the scratch.

**Comment by corpix** [comment]
I have closed #2, but I still think that urandom is better than `time.time()`(tracking this in #7), will move in this direction in future releases.
Thank you.
